File: test/Cartesian3d.cpp

#include <catch2/catch.hpp>
#include <xtensor/xrandom.hpp>

#define EQ(a,b) REQUIRE_THAT((a), Catch::WithinAbs((b), 1.e-12));

#include <GMatElastoPlasticQPot3d/Cartesian3d.h>

namespace GM = GMatElastoPlasticQPot3d::Cartesian3d;

template <class T, class S>
S A4_ddot_B2(const T& A, const S& B)
{
    S C;
    C.fill(0.0);

    for (size_t i = 0; i < 3; i++) {
        for (size_t j = 0; j < 3; j++) {
            for (size_t k = 0; k < 3; k++) {
                for (size_t l = 0; l < 3; l++) {
                    C(i, j) += A(i, j, k, l) * B(l, k);
                }
            }
        }
    }

    return C;
}

TEST_CASE("GMatElastoPlasticQPot3d::Cartesian3d", "Cartesian3d.h")
{

SECTION("Identity tensor: Id")
{
    GM::Tensor2 A = xt::random::randn<double>({3, 3});
    GM::Tensor2 I = GM::I2();
    GM::Tensor4 Id = GM::I4d();
    GM::Tensor4 Is = GM::I4s();
    A = A4_ddot_B2(Is, A);
    REQUIRE(xt::allclose(A4_ddot_B2(Id, A), A - GM::Hydrostatic(A) * I));
}

SECTION("Tangent (purely elastic response only)")
{
    double K = 12.3;
    double G = 45.6;

    GM::Tensor2 Eps = xt::random::randn<double>({3, 3});
    GM::Tensor2 Sig;
    GM::Tensor4 C;
    GM::Tensor4 Is = GM::I4s();
    Eps = A4_ddot_B2(Is, Eps);

    // Elastic
    {
        GM::Elastic mat(K, G);
        std::tie(Sig, C) = mat.Tangent(Eps);
        REQUIRE(xt::allclose(A4_ddot_B2(C, Eps), Sig));
    }

    // Cusp
    {
        GM::Cusp mat(K, G, {10000.0});
        std::tie(Sig, C) = mat.Tangent(Eps);
        REQUIRE(xt::allclose(A4_ddot_B2(C, Eps), Sig));
    }

    // Smooth
    {
        GM::Smooth mat(K, G, {10000.0});
        std::tie(Sig, C) = mat.Tangent(Eps);
        REQUIRE(xt::allclose(A4_ddot_B2(C, Eps), Sig));
    }
}

SECTION("Elasto-plastic response")
{
    double K = 12.3;
    double G = 45.6;

    GM::Tensor2 Eps;
    GM::Tensor2 Sig;

    double gamma = 0.02;
    double epsm = 0.12;

    Eps.fill(0.0);
    Eps(0,0) = Eps(1,1) = Eps(2,2) = epsm;
    Eps(0,1) = Eps(1,0) = gamma;

    // Elastic
    {
        GM::Elastic mat(K, G);
        Sig = mat.Stress(Eps);

        EQ(Sig(0,0), 3.0 * K * epsm);
        EQ(Sig(1,1), 3.0 * K * epsm);
        EQ(Sig(2,2), 3.0 * K * epsm);
        EQ(Sig(0,1), 2.0 * G * gamma);
        EQ(Sig(1,0), 2.0 * G * gamma);
    }

    // Cusp
    {
        GM::Cusp mat(K, G, {0.01, 0.03, 0.10});
        Sig = mat.Stress(Eps);

        EQ(Sig(0,0), 3.0 * K * epsm);
        EQ(Sig(1,1), 3.0 * K * epsm);
        EQ(Sig(2,2), 3.0 * K * epsm);
        EQ(Sig(0,1), 0.0);
        EQ(Sig(1,0), 0.0);
        EQ(mat.epsp(Eps), 0.02);
        REQUIRE(mat.find(Eps) == 1);
    }

    // Smooth
    {
        GM::Smooth mat(K, G, {0.01, 0.03, 0.10});
        Sig = mat.Stress(Eps);

        EQ(Sig(0,0), 3.0 * K * epsm);
        EQ(Sig(1,1), 3.0 * K * epsm);
        EQ(Sig(2,2), 3.0 * K * epsm);
        EQ(Sig(0,1), 0.0);
        EQ(Sig(1,0), 0.0);
        EQ(mat.epsp(Eps), 0.02);
        REQUIRE(mat.find(Eps) == 1);
    }

    // Matrix
    {
        size_t nelem = 3;
        size_t nip = 2;

        GM::Matrix mat(nelem, nip);

        {
            xt::xtensor<size_t,2> I = xt::zeros<size_t>({nelem, nip});
            xt::view(I, 0, xt::all()) = 1;
            mat.setElastic(I, K, G);
        }

        {
            xt::xtensor<size_t,2> I = xt::zeros<size_t>({nelem, nip});
            xt::xtensor<double,1> epsy = {0.01, 0.03, 0.10};
            xt::view(I, 1, xt::all()) = 1;
            mat.setCusp(I, K, G, epsy);
        }

        {
            xt::xtensor<size_t,2> I = xt::zeros<size_t>({nelem, nip});
            xt::xtensor<double,1> epsy = {0.01, 0.03, 0.10};
            xt::view(I, 2, xt::all()) = 1;
            mat.setCusp(I, K, G, epsy);
        }

        xt::xtensor<double,4> eps = xt::empty<double>({nelem, nip, 3ul, 3ul});

        for (size_t e = 0; e < nelem; ++e) {
            for (size_t q = 0; q < nip; ++q) {
                xt::view(eps, e, q) = Eps;
            }
        }

        auto sig = mat.Stress(eps);
        auto epsp = mat.Epsp(eps);

        for (size_t q = 0; q < nip; ++q) {

            EQ(sig(0,q,0,0), 3.0 * K * epsm);
            EQ(sig(0,q,1,1), 3.0 * K * epsm);
            EQ(sig(0,q,2,2), 3.0 * K * epsm);
            EQ(sig(0,q,0,1), 2.0 * G * gamma);
            EQ(sig(0,q,1,0), 2.0 * G * gamma);
            EQ(epsp(0,q), 0.0);

            EQ(sig(1,q,0,0), 3.0 * K * epsm);
            EQ(sig(1,q,1,1), 3.0 * K * epsm);
            EQ(sig(1,q,2,2), 3.0 * K * epsm);
            EQ(sig(1,q,0,1), 0.0);
            EQ(sig(1,q,1,0), 0.0);
            EQ(epsp(1,q), gamma);

            EQ(sig(2,q,0,0), 3.0 * K * epsm);
            EQ(sig(2,q,1,1), 3.0 * K * epsm);
            EQ(sig(2,q,2,2), 3.0 * K * epsm);
            EQ(sig(2,q,0,1), 0.0);
            EQ(sig(2,q,1,0), 0.0);
            EQ(epsp(2,q), gamma);
        }
    }
}

}

File: test/Cartesian3d.py

import GMatElastoPlasticQPot3d.Cartesian3d as GMat
import numpy as np

def EQ(a, b):
  assert np.abs(a-b) < 1.e-12

K = 12.3
G = 45.6

gamma = 0.02
epsm = 0.12

Eps = np.array(
    [[epsm, gamma, 0],
     [gamma, epsm, 0],
     [0, 0, epsm]])

# Elastic

mat = GMat.Elastic(K, G)
Sig = mat.Stress(Eps)

EQ(Sig[0,0], 3.0 * K * epsm)
EQ(Sig[1,1], 3.0 * K * epsm)
EQ(Sig[2,2], 3.0 * K * epsm)
EQ(Sig[0,1], 2.0 * G * gamma)
EQ(Sig[1,0], 2.0 * G * gamma)

# Cusp

mat = GMat.Cusp(K, G, [0.01, 0.03, 0.10])
Sig = mat.Stress(Eps)

EQ(Sig[0,0], 3.0 * K * epsm)
EQ(Sig[1,1], 3.0 * K * epsm)
EQ(Sig[2,2], 3.0 * K * epsm)
EQ(Sig[0,1], G * 0.0)
EQ(Sig[1,0], G * 0.0)
EQ(mat.epsp(Eps), 0.02)
EQ(mat.find(Eps), 1)

# Smooth

mat = GMat.Smooth(K, G, [0.01, 0.03, 0.10])
Sig = mat.Stress(Eps)

EQ(Sig[0,0], 3.0 * K * epsm)
EQ(Sig[1,1], 3.0 * K * epsm)
EQ(Sig[2,2], 3.0 * K * epsm)
EQ(Sig[0,1], G * 0.0)
EQ(Sig[1,0], G * 0.0)
EQ(mat.epsp(Eps), 0.02)
EQ(mat.find(Eps), 1)

# Matrix

nelem = 3
nip = 2
mat = GMat.Matrix(nelem, nip)

I = np.zeros([nelem, nip], dtype='int')
I[0,:] = 1
mat.setElastic(I, K, G)

I = np.zeros([nelem, nip], dtype='int')
I[1,:] = 1
mat.setCusp(I, K, G, [0.01, 0.03, 0.10])

I = np.zeros([nelem, nip], dtype='int')
I[2,:] = 1
mat.setSmooth(I, K, G, [0.01, 0.03, 0.10])

eps = np.zeros((nelem, nip, 3, 3))
for i in range(3):
    for j in range(3):
        eps[:, :, i, j] = Eps[i, j]

sig = mat.Stress(eps)
epsp = mat.Epsp(eps)

for q in range(nip):

    EQ(sig[0,q,0,0], 3.0 * K * epsm)
    EQ(sig[0,q,1,1], 3.0 * K * epsm)
    EQ(sig[0,q,2,2], 3.0 * K * epsm)
    EQ(sig[0,q,0,1], 2.0 * G * gamma)
    EQ(sig[0,q,0,1], 2.0 * G * gamma)
    EQ(epsp[0,q], 0.0)

    EQ(sig[1,q,0,0], 3.0 * K * epsm)
    EQ(sig[1,q,1,1], 3.0 * K * epsm)
    EQ(sig[1,q,2,2], 3.0 * K * epsm)
    EQ(sig[1,q,0,1], G * 0.0)
    EQ(sig[1,q,0,1], G * 0.0)
    EQ(epsp[1,q], gamma)

    EQ(sig[2,q,0,0], 3.0 * K * epsm)
    EQ(sig[2,q,1,1], 3.0 * K * epsm)
    EQ(sig[2,q,2,2], 3.0 * K * epsm)
    EQ(sig[2,q,0,1], G * 0.0)
    EQ(sig[2,q,0,1], G * 0.0)
    EQ(epsp[2,q], gamma)

print('All checks passed')
